umpost/stashmasters.py

```diff
@@ -4808,115 +4845,126 @@
-        3895: ["WOOD FLUX CARBON    (CASA-CNP)", "", "", "", ""],
-        3896: ["WOOD FLUX NITROGEN    (CASA-CNP)", "", "", "", ""],
-        3897: ["WOOD FLUX PHOSPHOR    (CASA-CNP)", "", "", "", ""],
-        3898: ["WOOD HARVEST CARBON1(CASA-CNP)", "", "", "", ""],
-        3899: ["WOOD HARVEST CARBON2(CASA-CNP)", "", "", "", ""],
-        3900: ["WOOD HARVEST CARBON3(CASA-CNP)", "", "", "", ""],
-        3901: ["WOOD HARVEST NITROG1(CASA-CNP)", "", "", "", ""],
-        3902: ["WOOD HARVEST NITROG2(CASA-CNP)", "", "", "", ""],
-        3903: ["WOOD HARVEST NITROG3(CASA-CNP)", "", "", "", ""],
-        3904: ["WOOD HARVEST PHOSPH1(CASA-CNP)", "", "", "", ""],
-        3905: ["WOOD HARVEST PHOSPH2(CASA-CNP)", "", "", "", ""],
-        3906: ["WOOD HARVEST PHOSPH3(CASA-CNP)", "", "", "", ""],
-        3907: ["WOOD RESPIRA CARBON1(CASA-CNP)", "", "", "", ""],
-        3908: ["WOOD RESPIRA CARBON2(CASA-CNP)", "", "", "", ""],
-        3909: ["WOOD RESPIRA CARBON3(CASA-CNP)", "", "", "", ""],
-        3910: ["WOOD RESPIRA NITROG1(CASA-CNP)", "", "", "", ""],
-        3911: ["WOOD RESPIRA NITROG2(CASA-CNP)", "", "", "", ""],
-        3912: ["WOOD RESPIRA NITROG3(CASA-CNP)", "", "", "", ""],
-        3913: ["WOOD RESPIRA PHOSPH1(CASA-CNP)", "", "", "", ""],
-        3914: ["WOOD RESPIRA PHOSPH2(CASA-CNP)", "", "", "", ""],
-        3915: ["WOOD RESPIRA PHOSPH3(CASA-CNP)", "", "", "", ""],
-        3916: ["THIN RATIO FOR FOREST (CASA-CNP)", "", "", "", ""],
-        3917: ["NITROGEN NET RELEASE (CASA-CNP)", "", "", "", ""],
-        3918: ["NITROGEN LEACHING (CASA-CNP)", "", "", "", ""],
-        3919: ["NITROGEN UPTAKE (CASA-CNP)", "", "", "", ""],
-        3920: ["NITROGEN LOSS (CASA-CNP)", "", "", "", ""],
-    },
+class STASHmaster(StrEnum):
+    """
+    Enumeration of the pre-set default STASHmasters.
+    """
+    CMIP6 = "cmip6"
+    ACCESS_CM2 = "access-cm2"
+    ACCESS_ESM1p5 = "access-esm1.5"
+    ACCESS_ESM1p6 = "access-esm1.6"
+    DEFAULT = CMIP6
 
-    # CM2 specific overwrites
-    MODEL_CM2: {
-        835: ["CABLE AQUIFER RELATIVE WATER (-)", "", "", "", ""],
-        889: ["PHOSPHORUS FERTILISER", "", "", "", ""],
-        890: ["SUM CARBON BALANCE ON TILES", "", "", "", ""],
-        891: ["SUM NITROGEN BALANCE ON TILES", "", "", "", ""],
-        892: ["SUM PHOSPHORUS BALANCE ON TILES", "", "", "", ""],
-        895: ["TEMPERATURE MET%TK FROM CABLE", "", "", "", ""],
-    }
+PRESET_STASHMASTERS = {
+    STASHmaster.CMIP6.value: CMIP6_DEFAULT_STASHMASTER,
+    STASHmaster.ACCESS_CM2.value: ACCESS_CM2_DEFAULT_STASHMASTER,
+    STASHmaster.ACCESS_ESM1p5.value: ACCESS_ESM1p5_DEFAULT_STASHMASTER,
+    STASHmaster.ACCESS_ESM1p6.value: ACCESS_ESM1p6_DEFAULT_STASHMASTER,
 }
 
-
 class StashVar:
-
     def __init__(self, code, model=None):
-        var = get_stashinfo(code, model)
+        if model is None:
+            model = STASHmaster.DEFAULT.value
+        self.stashmaster = PRESET_STASHMASTERS[model]
```

Comment:
I think it could be useful for a `StashVar` instance to keep track of which `STASHmaster` it has attached.
From a performance point of view it would not make any difference, but it would add potential uses in the future.

In the future, I think the whole `StashVar` should be revised (possibly renamed to `Stash` or `STASH`) and the `self.stashmaster` can be made to a read-only `@property` (potentially `@cachedproperty` for performance). We could also implement a setter if we want to allow changing `stashmaster` after the instance has been created, but I don't think there would be much use for that.

Reply:
But it's also true that, for the moment, we don't need this use, so it might make sense to turn it to a normal variable and then maybe turn it to an attribute/property in the future! 👍

Reply:
That's a fair point! Would another option be to store the name of the `stashmaster` that's being used in the object? e.g.
```python
self.model = model
```

Ok with either way though!

Reply:
I like the option to store only the name of the `stashmaster` for the moment! I'll implement it with `self.stashmaster = model`.

Reply:
I actually revised it slightly and fully changed the `model` keyword in `StashVar` to `stashmaster`.
Changes can be found in [43fdb17ac9dbcb226e28bbc8093dc99db51e623b](https://github.com/ACCESS-NRI/um2nc-standalone/pull/160/commits/43fdb17ac9dbcb226e28bbc8093dc99db51e623b)

Reply:
Those changes look good!